### PR TITLE
Add a requirements for the minimal memory_limit setting when installing

### PR DESCRIPTION
**Pull request by nstapelbroek**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | None
| License         | MIT

Hi! :raising_hand_man: 

When installing Sylius for the first time I got presented with an
```
[Symfony\Component\Debug\Exception\OutOfMemoryException]                                 
  Error: Allowed memory size of 134217728 bytes exhausted (tried to allocate 32768 bytes)
```
This PR should warn the user upfront when installing Sylius about any minimal memory requirements.
I'd love to hear your thoughts about this PR. I've just started my Sylius journey so I'm not that familair with the conventions yet. For example: Should we enforce a new memory_limit if `ini_set` is available?

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Resources/translations/messages.en.yml [hunk 1]
```diff
@@ -202,6 +202,7 @@ sylius:
                 help: Timezone %timezone% is deprecated. Fix your php.ini file (list of deprecated timezones http://php.net/manual/en/timezones.others.php).
             version: PHP version
             version_recommended: Recommended PHP version
+            memory_limit: memory_limit
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Resources/translations/messages.en.yml]: You should add translation key only in this file with the proper message and revert adding this line in other `messages.*.yml` files

**Comment by pamil** [comment]
I don't think we can estimate how much memory is really needed, further PHP versions optimises its memory usage so we might end up with installation failing because of this requirement only.

**Comment by nstapelbroek** [comment]
@pamil good point, would you like me to change the requirement into a "warning"? I haven't tried this one out on multiple systems. But form here it looks like the default of 128m is just not enough for running the installation.

File: src/Sylius/Bundle/CoreBundle/Resources/translations/messages.en.yml (hunk 1 above)
Comment (nstapelbroek) [reply src/Sylius/Bundle/CoreBundle/Resources/translations/messages.en.yml]: Alright! I'll fix this tonight

**Comment by nstapelbroek** [comment]
Implemented PR feedback and squashed the commits. Changes:
- Instead of a requirement, the assertion is now a check and will not block the installation
- Removed duplicate translations
- Added a help message

**Comment by pjedrzejewski** [comment]
@nstapelbroek Nice, looks like a good change to me! Could you have a look at the failing build, please? Thanks!

File: src/Sylius/Bundle/CoreBundle/Installer/Requirement/SettingsRequirements.php [hunk 2]
```diff
@@ -47,7 +48,12 @@ public function __construct(TranslatorInterface $translator)
                 !$this->isOn('session.auto_start'),
                 false
             ))
-        ;
+            ->add(new Requirement(
+                $translator->trans('sylius.installer.settings.memory_limit.header', []),
+                $this->hasSufficientMemory(),
+                false,
+                $translator->trans('sylius.installer.settings.memory_limit.help', [])
+            ));
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Installer/Requirement/SettingsRequirements.php]: `;` should be in the next line

File: src/Sylius/Bundle/CoreBundle/Installer/Requirement/SettingsRequirements.php [hunk 3]
```diff
@@ -61,4 +67,47 @@ private function isOn(string $key): bool
...
+            case 'g':
+                $value *= 1024;
+            // no break (cumulative multiplier)
+            case 'm':
+                $value *= 1024;
+            // no break (cumulative multiplier)
+            case 'k':
+                $value *= 1024;
+        }
+        return $value;
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Installer/Requirement/SettingsRequirements.php]: Missing blank line before `return`

File: src/Sylius/Bundle/CoreBundle/Installer/Requirement/SettingsRequirements.php [hunk 4]
```diff
@@ -61,4 +67,47 @@ private function isOn(string $key): bool
...
+     *
+     * @see https://github.com/composer/composer/blob/master/bin/composer#L26
+     *
+     * @param $value
+     * @return int
+     */
+    private function getMemoryInBytes($value)
+    {
+        $unit = strtolower(substr($value, -1, 1));
+        $value = (int)$value;
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Installer/Requirement/SettingsRequirements.php]: Missing space after `(int)`

File: src/Sylius/Bundle/CoreBundle/Installer/Requirement/SettingsRequirements.php [hunk 5]
```diff
@@ -61,4 +67,47 @@ private function isOn(string $key): bool
...
+        return false;
+    }
+
+    /**
+     * Converts a memory_limit value to amount of bytes
+     *
+     * @see https://github.com/composer/composer/blob/master/bin/composer#L26
+     *
+     * @param $value
+     * @return int
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Installer/Requirement/SettingsRequirements.php]: Missing blank line before `@return`

File: src/Sylius/Bundle/CoreBundle/Installer/Requirement/SettingsRequirements.php [hunk 6]
```diff
@@ -61,4 +67,47 @@ private function isOn(string $key): bool
...
+     */
+    private function getMemoryInBytes($value)
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Installer/Requirement/SettingsRequirements.php]: `private function getMemoryInBytes(string $value): int`?

**Comment by pamil** [comment]
I agree it's a nice change, but still, we can't estimate how much memory the installation would take. I've just made a few installs and these are the results:

PHP version | Environment | Options | Peak memory usage
-------------|--------------|---------|----------------
PHP 7.1 | dev | sample data | 174 MiB
PHP 7.1 | dev | no sample data | 146 MiB

Even when using `dev` environment, every installation was below those required 256 MiB (and it can go even lower with OPCache enabled for CLI or next PHP versions).

**Comment by stefandoorn** [comment]
@pamil How does this memory usage go when doing a full `composer update` on an installation?

**Comment by pamil** [comment]
@stefandoorn this one relates to `sylius:install` command only, which happens after Composer installation / update. Last time I've checked, `composer update` took ~1.2 GiB on PHP 7.1.

**Comment by pamil** [comment]
I've just done a `composer update` on `Sylius/Sylius` with PHP 7.1 and peak memory usage was 700 MiB.

**Comment by nstapelbroek** [comment]
Thanks, all. If I understand correctly I only need to apply the requested changes of @GSadee ?
[Composer sets their own memory_limit](https://github.com/composer/composer/blob/eb5495b8e40811fc637e64bf0971a2e2349400ad/bin/composer#L46) if `ini_set` is available. Might this be a solution we could use here?

**Comment by nstapelbroek** [comment]
Hiya :raising_hand_man: , any updates on this? Do I need to do something to get this merged?

**Comment by pamil** [comment]
Hey! I'm sorry to keep it stale. As I've mentioned in a comment above, Sylius can be installed using only 146MiB memory, so 256MiB requirement would prevent it from a successful installation and that's not what we want.

I guess it's perfectly fine to try to change the memory limit to 256MiB if it's below that level, just as [Composer does](https://github.com/composer/composer/blob/master/bin/composer#L45-L47), but I'd rather not prevent end users from trying to install Sylius just because of a lower memory limit.

**Pull request closed without merging**